src/compiler/checker.ts

```diff
@@ -25375,47 +25376,101 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function isMatchingReference(source: Node, target: Node): boolean {
-        switch (target.kind) {
-            case SyntaxKind.ParenthesizedExpression:
-            case SyntaxKind.NonNullExpression:
-                return isMatchingReference(source, (target as NonNullExpression | ParenthesizedExpression).expression);
-            case SyntaxKind.BinaryExpression:
-                return (isAssignmentExpression(target) && isMatchingReference(source, target.left)) ||
-                    (isBinaryExpression(target) && target.operatorToken.kind === SyntaxKind.CommaToken && isMatchingReference(source, target.right));
-        }
-        switch (source.kind) {
+        const keySource = getReferenceCacheKey(source);
+        const keyTarget = getReferenceCacheKey(target);
+        return keySource !== undefined && keyTarget !== undefined && keySource === keyTarget;
+    }
+
+    /**
+     * @returns A cache key that looks like `1234."field"."\"prop\".\"name\""` or `undefined` if the node can't be a reference

```

Comment:
While this is somewhat resilient against antagonistic field names, the escaping is potentially costly, and I'd like to try to avoid it if this shows promise. A less readable key format, like `1234.23.57` might be better (where `23` and `57` are indexes into a known-field-name-strings holding array). Still, I've cached the results, so it's not like we repeat the work, so savings by making this less readable aren't likely to be huge.